var/spack/repos/builtin/packages/openfoam/package.py

```diff
@@ -0,0 +1,358 @@
+
+    version('1612', 'ca02c491369150ab127cbb88ec60fbdf',
+        url = base_url+'/v1612+/OpenFOAM-v1612+.tgz')
+
+    variant('scotch', default=True,
+        description='With scotch for decomposition')
+    variant('ptscotch', default=True,
+        description='With pt-scotch for decomposition')
+    variant('metis', default=True,
+        description='With metis for decomposition')
+    ## Not yet working properly
+    ##    variant('parmgridgen', default=True,  ## Not yet working properly
+    ##        description='With parmgridgen support')
+    variant(
+        'source', default=True,
+        description='Install library/application sources and tutorials')
+
+    variant('paraview', default=True,
+        description='Build paraview plugins and runtime post-processing')
+
+    supported_compilers = {'clang': 'Clang', 'gcc': 'Gcc', 'intel': 'Icc'}
+
+    depends_on('mpi')
+    depends_on('flex')
+    depends_on('zlib')
+    depends_on('fftw')
+    depends_on('boost')
+    depends_on('cgal')
+    depends_on('cmake', type='build')
+
+    depends_on('scotch ~ metis', when='~ptscotch+scotch')
+    depends_on('scotch ~ metis + mpi', when='+ptscotch')
+    depends_on('metis@5:', when='+metis')
+    depends_on('parmgridgen', when='+parmgridgen')
+
+    # For OpenFOAM plugins and run-time post-processing this should really just
+    # be 'paraview +plugins' but that resolves poorly.
+    # Instead place the preferred variants "+plugins +qt"
+    # in the ~/.spack/packages.yaml
+    depends_on('paraview', when='+paraview')
+
+
+    # General patches
+    patch('openfoam-build.patch')
+    patch('openfoam-site.patch')
+
+    # Things already patched upstream
+    patch('openfoam-bin-1612.patch', when='@1612')
+    patch('openfoam-etc-1612.patch', when='@1612')
+    patch('scotch-metis-lib-1612.patch', when='@1612')
+
+
+    def get_openfoam_arch(self):
+        """Return equivalent to WM_ARCH, WM_COMPILER"""
+        (sysname, nodename, release, version, machine) = os.uname()
+
+        if self.compiler.name not in self.supported_compilers:
+            raise RuntimeError('{0} is not a supported compiler \
+            to compile OpenFOAM'.format(self.compiler.name))
+
+        wm_compiler = self.supported_compilers[self.compiler.name]
+        if sysname == 'Linux':
+            wm_arch = 'linux'
+            # Only suppport 64-bit on x86_64
+            if machine == 'x86_64':
+                wm_arch = 'linux64'
+        else:
+            raise RuntimeError('{0} {1} is not a \
+            supported architecture'.format(sysname, machine))
+
+        return (wm_arch, wm_compiler)
+
+
+    ## Needed in older OpenFOAM versions (prior to 1612), but now has no effect
+    ##
+    ##    def setup_environment(self, spack_env, run_env):
+    ##        spack_env.set('FOAM_INST_DIR', os.path.abspath(self.stage.path))
+    ##        run_env.set('FOAM_INST_DIR', self.prefix)
+
+
+    def patch(self):
```

Comment:
As a docstring, please add the justification and reasoning behind these patches.